internal/pkg/deploy/upload/customresource/customresource.go

```diff
@@ -0,0 +1,185 @@
+		customDomainFnName:  customDomainAppRunnerFilePath,
+	})
+}
+
+// LBWS returns the custom resources for a load-balanced web service.
+func LBWS(fs template.Reader) ([]CustomResource, error) {
+	return buildCustomResources(fs, map[string]string{
+		dynamicDesiredCountFnName: desiredCountDelegationFilePath,
+		envControllerFnName:       envControllerFilePath,
+		rulePriorityFnName:        albRulePriorityGeneratorFilePath,
+		nlbCustomDomainFnName:     nlbCustomDomainFilePath,
+		nlbCertValidatorFnName:    nlbCertValidatorFilePath,
+	})
+}
+
+// Worker returns the custom resources for a worker service.
+func Worker(fs template.Reader) ([]CustomResource, error) {
+	return buildCustomResources(fs, map[string]string{
+		dynamicDesiredCountFnName: desiredCountDelegationFilePath,
+		backlogPerTaskFnName:      backlogPerTaskCalculatorFilePath,
+		envControllerFnName:       envControllerFilePath,
+	})
+}
+
+// Backend returns the custom resources for a backend service.
+func Backend(fs template.Reader) ([]CustomResource, error) {
+	return buildCustomResources(fs, map[string]string{
+		dynamicDesiredCountFnName: desiredCountDelegationFilePath,
+		rulePriorityFnName:        albRulePriorityGeneratorFilePath,
+		envControllerFnName:       envControllerFilePath,
+	})
+}
+
+// ScheduledJob returns the custom resources for a scheduled job.
+func ScheduledJob(fs template.Reader) ([]CustomResource, error) {
+	return buildCustomResources(fs, map[string]string{
+		envControllerFnName: envControllerFilePath,
+	})
+}
+
+// Env returns the custom resources for an environment.
+func Env(fs template.Reader) ([]CustomResource, error) {
+	return buildCustomResources(fs, map[string]string{
+		certValidationFnName: dnsCertValidationFilePath,
+		customDomainFnName:   customDomainFilePath,
+		dnsDelegationFnName:  dnsDelegationFilePath,
+	})
+}
+
+// UploadFunc is the function signature to upload contents under a key within a S3 bucket.
+type UploadFunc func(key string, contents io.Reader) (url string, err error)
+
+// Upload zips all the Files for each CustomResource and uploads the zip files individually to S3.
+// Returns a map of the CustomResource FunctionName to the S3 URL where the zip file is stored.
+func Upload(upload UploadFunc, crs []CustomResource) (map[string]string, error) {
+	urls := make(map[string]string)
+	for _, cr := range crs {
+		zipFile, err := cr.Zip()
+		if err != nil {
+			return nil, err
+		}
+		out, err := io.ReadAll(zipFile)
+		if err != nil {
+			return nil, fmt.Errorf("read content of zip file for custom resource %q: %v", cr.FunctionName(), err)
+		}
+		url, err := upload(artifactpath.CustomResource(strings.ToLower(cr.FunctionName()), out), zipFile)
+		if err != nil {
+			return nil, fmt.Errorf("upload custom resource %q: %w", cr.FunctionName(), err)
+		}
+		urls[cr.FunctionName()] = url
+	}
+	return urls, nil
+}
+
+func buildCustomResources(fs template.Reader, pathForFn map[string]string) ([]CustomResource, error) {
+	var crs []CustomResource
```

Comment:
tiny little nit: perhaps in this case we could allocate the capacity here since it's known!